src/unif.ml

```diff
@@ -102,28 +116,62 @@ let unify_modulo : term -> term -> bool = fun a b ->
-        let (a,sa) = whnf_stk a [] in
-        let (b,sb) = whnf_stk b [] in
-        let rec sync acc la lb =
-          match (la, lb) with
-          | ([]   , []   ) -> (a, b, acc)
-          | (a::la, b::lb) -> sync ((!a,!b)::acc) la lb
-          | (la   , []   ) -> (to_term a (List.rev la), b, acc)
-          | ([]   , lb   ) -> (a, to_term b (List.rev lb), acc)
-        in
-        let (a,b,l) = sync l (List.rev sa) (List.rev sb) in
-        match (a, b) with
-        | (a          , b          ) when unify a b -> unify_modulo l
-        | (Abst(aa,ba), Abst(ab,bb)) ->
-            let (_,ba,bb) = Bindlib.unbind2 mkfree ba bb in
-            unify_modulo ((aa,ab)::(ba,bb)::l)
-        | (Prod(aa,ba), Prod(ab,bb)) ->
-            let (_,ba,bb) = Bindlib.unbind2 mkfree ba bb in
-            unify_modulo ((aa,ab)::(ba,bb)::l)
-        | (a          , b          ) ->
+    | (a,b)::l                ->
+    let a = whnf a in
+    let b = whnf b in
+    let rec get_args args a b =
+      match (unfold a, unfold b) with
+      | (Appl(t1,u1), Appl(t2,u2)) -> get_args ((u1,u2)::args) t1 t2
+      | (Meta(u1,e1), Meta(u2,e2)) when u1 == u2 -> (a,b,args)
+      | (Meta(m,e)  , b          ) when instantiate m e b -> get_args args a b
+      | (a          , Meta(m,e)  ) when instantiate m e a -> get_args args a b
+      | (_          , _          ) -> (a,b,args)
+    in
+    let unif =
+      match !constraints with
+      | None -> (fun a b -> unify_modulo [(a,b)])
+      | _    -> unify
+    in
+    let (a,b,args) = get_args [] a b in
+    let args = List.rev args in
+    match (unfold a, unfold b) with
+    | (Wild         , _            )
+    | (_            , Wild         )
+    | (ITag(_)      , _            )
+    | (_            , ITag(_)      )
+    | (Type         , Type         ) -> assert (args = []); unify_modulo l
+    | (Kind         , Kind         ) -> assert (args = []); unify_modulo l
+    | (Prod(a1,b1)  , Prod(a2,b2)  ) ->
+        assert (args = []);
+        let (_,b1,b2) = Bindlib.unbind2 mkfree b1 b2 in
+        unify_modulo ((a1,a2)::(b1,b2)::l)
+    | (Abst(a1,t1)  , Abst(a2,t2)  ) ->
+        assert (args = []);
+        let (_,t1,t2) = Bindlib.unbind2 mkfree t1 t2 in
+        unify_modulo ((a1,a2)::(t1,t2)::l)
+    | (Vari(x1)     , Vari(x2)     ) when Bindlib.eq_vars x1 x2 ->
+        List.for_all (fun (a,b) -> unif a b || add_constraint a b) args &&
+        unify_modulo l
+    | (Symb(Sym(s1)), Symb(Sym(s2))) ->
+        s1 == s2 && unify_modulo (args @ l)
+    | (Symb(Def(s1)), Symb(Def(s2))) when s1 == s2 && args = [] ->
+        unify_modulo l
+    | (Meta(u1,e1)  , Meta(u2,e2)  ) ->
+        u1 == u2 &&
+        Array.for_all2 (fun a b -> unif a b || add_constraint a b) e1 e2 &&
+        List.for_all (fun (a,b) -> unif a b || add_constraint a b) args &&
+        unify_modulo l
+    | (a            , b            ) ->
+        match !constraints with
+        | None ->
+            unify a b &&
```

Comment:
Why are the possibilities for [unify a b] to hold here?

Moreover, if the Appl case in unify is not fixed, injectivity may be implicitly used here again.

Reply:
There is the definable constant case (with non-empty list of arguments).